lib/internal/modules/esm/loader.js

```diff
@@ -515,9 +522,24 @@ function createModuleLoader(useCustomLoadersIfPresent = true) {
       !require('internal/modules/esm/utils').isLoaderWorker()) {
     const userLoaderPaths = getOptionValue('--experimental-loader');
     if (userLoaderPaths.length > 0) {
-      if (!emittedExperimentalWarning) {
-        emitExperimentalWarning('Custom ESM Loaders');
-        emittedExperimentalWarning = true;
+      if (!emittedLoaderFlagWarning) {
+        const readableURIEncode = (string) => ArrayPrototypeReduce(
+          [
+            [/'/g, '%27'], // We need to URL-encode the single quote as it's the delimiter for the --import flag.
+            [/%22/g, '"'], // We can decode the double quotes to improve readability.
+            [/%2F/ig, '/'], // We can decode the slashes to improve readability.
+          ],
+          (str, { 0: regex, 1: replacement }) => RegExpPrototypeSymbolReplace(hardenRegExp(regex), str, replacement),
+          encodeURIComponent(string));
+        process.emitWarning(
+          '`--experimental-loader` may be removed in the future; instead use `register()`:\n' +
+          `--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; ${ArrayPrototypeJoin(
+            ArrayPrototypeMap(userLoaderPaths, (loader) => `register(${readableURIEncode(JSONStringify(loader))}, pathToFileURL("./"))`),
```

Comment:
This data URL example feels a little misleading, when a separate file will be the main use case at this point. Is there a reason it has to be using this framing over a more recommended example?

Reply:
This is a migration path, for packages that are currently expecting to be used via `--loader` / `--experimental-loader`. This warning is only printed when those flags specifically are passed. Any packages not already recommending users use `--import` might not have the appropriate file to be passed to `--import`.

Reply:
Thanks for explaining, that makes sense.